compiler/rustc_trait_selection/src/solve/inherent_projection.rs

```diff
@@ -0,0 +1,38 @@
+use rustc_middle::traits::solve::{Certainty, Goal, QueryResult};
+use rustc_middle::ty;
+
+use super::EvalCtxt;
+
+impl<'tcx> EvalCtxt<'_, 'tcx> {
+    pub(super) fn normalize_inherent_associated_type(
+        &mut self,
+        goal: Goal<'tcx, ty::ProjectionPredicate<'tcx>>,
+    ) -> QueryResult<'tcx> {
+        let tcx = self.tcx();
+        let inherent = goal.predicate.projection_ty;
+        let expected = goal.predicate.term.ty().expect("inherent consts are treated separately");
+
+        let impl_def_id = tcx.parent(inherent.def_id);
+        let impl_substs = self.fresh_substs_for_item(impl_def_id);
+
+        // Equate impl header and add impl where clauses
+        self.eq(
+            goal.param_env,
+            inherent.self_ty(),
+            tcx.type_of(impl_def_id).subst(tcx, impl_substs),
+        )?;
+        self.add_goals(
+            tcx.predicates_of(impl_def_id)
```

Comment:
Here, just like the normal projection code, https://github.com/rust-lang/rust/blob/4dbc7e3092fce6f3d5712733e5206226183dbe1e/compiler/rustc_trait_selection/src/solve/trait_goals.rs#L70-L76 we only register the impl's where clauses. the GAT's own where clauses are guaranteed by well-formedness.

Not exactly sure if we should be registering both. Maybe it affects inference?? :shrug:

Reply:
please also check GAT where clauses for this and ordinary trait impl candidates https://github.com/rust-lang/trait-system-refactor-initiative/issues/44